### PR TITLE
fix clearInterval bug and ShadowDOM  bug for ie 10 and 11

### PR DESCRIPTION
**Pull request by xinhailishi**
##### Checklist

- [x] commit message follows commit guidelines

##### Description of change

- close https://github.com/umijs/qiankun/issues/1489

### CODE REVIEW

**Comment by kuitos** [comment]
Could you pls revert the unrelated changes?

**Comment by xinhailishi** [comment]
> Could you pls revert the unrelated changes?

Could you tell me what is  the unrelated change?    I just think all  is need.

**Comment by xinhailishi** [comment]
> Could you pls revert the unrelated changes?

I update the linked  issue #1489   @kuitos   .
now there is no unrelated change.

File: src/sandbox/patchers/interval.ts [hunk 1]
```diff
@@ -7,14 +7,17 @@
...
+window.rawWindowClearInterval = window.clearInterval;
 
 export default function patch(global: Window) {
   let intervals: number[] = [];
 
   global.clearInterval = (intervalId: number) => {
     intervals = intervals.filter((id) => id !== intervalId);
-    return rawWindowClearInterval(intervalId);
+    // @ts-ignore
+    return window.rawWindowClearInterval(intervalId);
```
Comment (kuitos) [inline src/sandbox/patchers/interval.ts:20]: I guess use `rawWindowClearInterval.call(window, intervalId)` could solve the compatible issue, rather than bind a new global variable on window.

File: src/sandbox/snapshotSandbox.ts [hunk 2]
```diff
@@ -8,9 +8,10 @@ import { SandBoxType } from '../interfaces';
 function iter(obj: typeof window, callbackFn: (prop: any) => void) {
   // eslint-disable-next-line guard-for-in, no-restricted-syntax
   for (const prop in obj) {
-    if (obj.hasOwnProperty(prop)) {
-      callbackFn(prop);
-    }
+    // ie10,11下默认hasOwnProperty不包括clearInterval等
+    // if (obj.hasOwnProperty(prop)) {
```
Comment (kuitos) [inline src/sandbox/snapshotSandbox.ts:12]: Why need change this method? It is consistent both in active or inactive period, while iterator can not catch clearInterval property on window.

**Review by kuitos — CHANGES_REQUESTED** [review]

File: src/sandbox/snapshotSandbox.ts (hunk 2 above)
Comment (xinhailishi) [reply src/sandbox/snapshotSandbox.ts:12]: run this  `window.hasOwnProperty('clearInterval')`   under  ie10,11.  Then you can see  'false',but  'true'   under chrome.   It could lead to  that snapshotSandbox can not revert under  ie10,11.

Comment (xinhailishi) [reply src/sandbox/snapshotSandbox.ts:12]: @kuitos  it  is different in inactive and in active,because   you rewrite  cleartInterval   in  interval.ts   `global.clearInterval=` in active. Once you rewrite clearInterval,`window.hasOwnProperty('clearInterval')` will be `true` .  This causes differences between in inactive and in active.

File: src/sandbox/patchers/interval.ts (hunk 1 above)
Comment (xinhailishi) [reply src/sandbox/patchers/interval.ts:20]: > I guess use `rawWindowClearInterval.call(window, intervalId)` could solve the compatible issue, rather than bind a new global variable on window.

@kuitos  you are right.   `rawWindowClearInterval.call(window, intervalId)` works too.

Comment (xinhailishi) [reply src/sandbox/patchers/interval.ts]: > I guess use `rawWindowClearInterval.call(window, intervalId)` could solve the compatible issue, rather than bind a new global variable on window.

I fix it

File: src/loader.ts [hunk 3]
```diff
@@ -320,12 +320,13 @@ export async function loadApp<T extends ObjectType>(
...
-    lifeCycles,
-    (v1, v2) => concat(v1 ?? [], v2 ?? []),
-  );
+  const {
+    beforeUnmount = [],
+    afterUnmount = [],
+    afterMount = [],
+    beforeMount = [],
+    beforeLoad = [],
+  } = mergeWith({}, getAddOns(global, assetPublicPath), lifeCycles, (v1, v2) => concat(v1 ?? [], v2 ?? []));
```
Comment (gongshun) [inline src/loader.ts:329]: 代码格式不要修改，需要还原回去

File: src/loader.ts [hunk 4]
```diff
@@ -338,11 +339,8 @@ export async function loadApp<T extends ObjectType>(
     sandboxContainer?.instance?.latestSetProp,
   );
 
-  const {
-    onGlobalStateChange,
-    setGlobalState,
-    offGlobalStateChange,
-  }: Record<string, CallableFunction> = getMicroAppStateActions(appInstanceId);
+  const { onGlobalStateChange, setGlobalState, offGlobalStateChange }: Record<string, CallableFunction> =
+    getMicroAppStateActions(appInstanceId);
```
Comment (gongshun) [inline src/loader.ts:343]: 这里也是，代码格式不要修改，需要还原回去

File: src/sandbox/patchers/interval.ts [hunk 5]
```diff
@@ -7,14 +7,17 @@
 import { noop } from 'lodash';
 
 const rawWindowInterval = window.setInterval;
+// 兼容ie10、11下，非全局clearInterval,报调用的对象无效，
```
Comment (kuitos) [inline src/sandbox/patchers/interval.ts:10]:
```suggestion
```

File: src/sandbox/patchers/interval.ts [hunk 6]
```diff
@@ -7,14 +7,17 @@
...
+// @ts-ignore
```
Comment (kuitos) [inline src/sandbox/patchers/interval.ts:11]:
```suggestion
// @ts-ignore
```

File: src/sandbox/patchers/interval.ts [hunk 7]
```diff
@@ -7,14 +7,17 @@
...
 const rawWindowClearInterval = window.clearInterval;
 
 export default function patch(global: Window) {
   let intervals: number[] = [];
 
   global.clearInterval = (intervalId: number) => {
     intervals = intervals.filter((id) => id !== intervalId);
-    return rawWindowClearInterval(intervalId);
+    // @ts-ignore
```
Comment (kuitos) [inline src/sandbox/patchers/interval.ts:19]:
```suggestion
    // @ts-ignore
```

File: src/sandbox/patchers/interval.ts [hunk 8]
```diff
@@ -26,6 +29,7 @@ export default function patch(global: Window) {
   return function free() {
     intervals.forEach((id) => global.clearInterval(id));
     global.setInterval = rawWindowInterval;
+    // @ts-ignore
```
Comment (kuitos) [inline src/sandbox/patchers/interval.ts:32]:
```suggestion
```

File: src/sandbox/snapshotSandbox.ts (hunk 2 above)
Comment (kuitos) [inline src/sandbox/snapshotSandbox.ts:12]:
```suggestion
    if (obj.hasOwnProperty(prop) || prop === 'clearInterval') {
```

File: src/sandbox/snapshotSandbox.ts [hunk 9]
```diff
@@ -8,9 +8,10 @@ import { SandBoxType } from '../interfaces';
...
+    callbackFn(prop);
```
Comment (kuitos) [inline src/sandbox/snapshotSandbox.ts:13]:
```suggestion
      callbackFn(prop);
```

File: src/sandbox/snapshotSandbox.ts [hunk 10]
```diff
@@ -8,9 +8,10 @@ import { SandBoxType } from '../interfaces';
...
+    // }
```
Comment (kuitos) [inline src/sandbox/snapshotSandbox.ts:14]:
```suggestion
    }
```

File: src/loader.ts (hunk 3 above)
Comment (kuitos) [reply src/loader.ts:329]: 这个应该是 prettier 自动格式化的

**Review by kuitos — CHANGES_REQUESTED** [review]

File: src/sandbox/snapshotSandbox.ts [hunk 11]
```diff
@@ -8,7 +8,8 @@ import { SandBoxType } from '../interfaces';
 function iter(obj: typeof window, callbackFn: (prop: any) => void) {
   // eslint-disable-next-line guard-for-in, no-restricted-syntax
   for (const prop in obj) {
-    if (obj.hasOwnProperty(prop)) {
+    // ie10,11下默认hasOwnProperty不包括clearInterval等
```
Comment (kuitos) [inline src/sandbox/snapshotSandbox.ts:11]:
```suggestion
    // patch for clearInterval for compatible reason, see #1490 
```

File: src/loader.ts [hunk 12]
```diff
@@ -135,7 +135,7 @@ function getAppWrapperGetter(
       `[qiankun] Wrapper element for ${appName} with instance ${appInstanceId} is not existed!`,
     );
 
-    if (strictStyleIsolation) {
+    if (strictStyleIsolation && supportShadowDOM) {
```
Comment (kuitos) [inline src/loader.ts:138]:
```suggestion
    if (strictStyleIsolation) {
```

Do not fix issues more than one in a pr, split the other fixed into another prs

**Review by kuitos — CHANGES_REQUESTED** [review]

**Review by kuitos — APPROVED** [review]
Pls submit another pr to fix shadow dom compatible issue

File: src/sandbox/snapshotSandbox.ts [hunk 13]
```diff
@@ -8,7 +8,8 @@ import { SandBoxType } from '../interfaces';
 function iter(obj: typeof window, callbackFn: (prop: any) => void) {
   // eslint-disable-next-line guard-for-in, no-restricted-syntax
   for (const prop in obj) {
-    if (obj.hasOwnProperty(prop)) {
+    // patch for clearInterval for compatible reason, see #1490
```
Comment (kuitos) [inline src/sandbox/snapshotSandbox.ts:11]:
```suggestion
    // patch for clearInterval for compatible reason, see #1490
```

File: src/sandbox/patchers/interval.ts [hunk 14]
```diff
@@ -14,7 +14,7 @@ export default function patch(global: Window) {
 
   global.clearInterval = (intervalId: number) => {
     intervals = intervals.filter((id) => id !== intervalId);
-    return rawWindowClearInterval(intervalId);
+    return rawWindowClearInterval.call(window, intervalId);
```
Comment (kuitos) [inline src/sandbox/patchers/interval.ts:17]:
```suggestion
    return rawWindowClearInterval.call(window, intervalId as any);
```

**Comment by xinhailishi** [comment]
> Pls submit another pr to fix shadow dom compatible issue

ok @kuitos

**Comment by HardyDou** [comment]
现在还在用qiankun 吗？用的怎么样？ @xinhailishi
